### PR TITLE
Classes Rework P.1

### PR DESCRIPTION
**Pull request by LoneWolfHT**
* New class select GUI
* New ctf_gui API
* Rework for scouts
* fix damage cobble not damaging shooters

### CODE REVIEW

File: mods/apis/ctf_gui/dev.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+	assert(file, error)
+
+	file:write(formspec)
+
+	file:close()
+
+	local function interval()
+		if formspec:sub(1, 3) == "[f]" then
+			local result, form = pcall(loadstring(formspec:sub(4)), formcontext)
```
Comment (appgurueu) [inline mods/apis/ctf_gui/dev.lua:17]: There is absolutely no need to use loadstring here?

Comment (LoneWolfHT) [reply mods/apis/ctf_gui/dev.lua:17]: Does `pcall` run strings as lua code?

Comment (LoneWolfHT) [reply mods/apis/ctf_gui/dev.lua:17]: [f] indicates you are writing lua code that returns the formspec string

Comment (appgurueu) [reply mods/apis/ctf_gui/dev.lua:17]: > Does `pcall` run strings as lua code?

`pcall` runs a function. When you call `loadstring`, you are "compiling" Lua code.

**Comment by appgurueu** [comment]
I recommend you to have a look at [Epidermis formspec implementation](https://github.com/appgurueu/epidermis/blob/master/formspec.lua) for inspiration; it's simple yet effective way to show a formspec to a player with a callback attached to it.

Note: There is no need to use an explicit `context[playername]` table if you use my approach; all upvalues to the callback will simply be the "context". This is both faster and more comfortable to use.

File: mods/apis/ctf_gui/dev.lua (hunk 1 above)
Comment (appgurueu) [reply mods/apis/ctf_gui/dev.lua:17]: > [f] indicates you are writing lua code that returns the formspec string

Then you should instead allow passing/storing a function that does this. Passing Lua code around as prefixed strings and compiling it upon execution is bad for both performance and code quality.

Comment (LoneWolfHT) [reply mods/apis/ctf_gui/dev.lua:17]: This is for live formspec editing, not release use (see the non _dev function for that).
I'll make this clearer.

Comment (appgurueu) [reply mods/apis/ctf_gui/dev.lua:17]: > This is for live formspec editing, not release use (see the non _dev function for that). I'll make this clearer.

I wouldn't want this for live formspec editing either. You'll have some kind of Lua prompt and should just be entering the function there. Having to put it in a string is annoying.

Comment (LoneWolfHT) [reply mods/apis/ctf_gui/dev.lua:17]: Down below it creates `file_edit.txt` in the world directory. It loads that file as a string every second, converts it to lua code if that's what it says it is, and runs that lua code for the formspec, which it then displays

**Comment by LoneWolfHT** [comment]
*The one time I was too lazy to diff my changes against the main remote*

**Pull request merged**